.circleci/wait-until-previous-builds-complete.sh

```diff
@@ -0,0 +1,165 @@
+    exit 1
+  fi
+}
+
+post() {
+  local url="${api_root}${1}"
+  printf "curl POST ${url}\n" >/dev/tty
+  curl -X POST \
+    -H "Content-Type: application/json" \
+    -H "Circle-Token: ${CIRCLECI_TOKEN}" \
+    -d "$2" \
+    "${url}"
+}
+
+get() {
+  local url="${api_root}${1}"
+  printf "curl GET ${url}\n" >/dev/tty
+  curl -X GET -sS \
+    -H "Content-Type: application/json" \
+    -H "Circle-Token: ${CIRCLECI_TOKEN}" \
+    "${url}"
+}
+
+# Get list of recently built pipelines. Save results to json file.
+pipeline_json="/tmp/master_branch_pipelines.json"
+fetch_pipeline_ids() {
+  local get_path="pipeline?org-slug=${project_slug}"
+  local get_result=$(get "${get_path}")
+  echo "${get_result}" | jq "[.items[] | select(.vcs.branch==\"master\")][] | {created_at: .created_at, id: .id, number: .number}" > "${pipeline_json}"
+  printf "Found following pipelines on ${branch} branch:\n"
+  cat ${pipeline_json}
+  printf "\n"
+}
+
+fetch_pipeline_number() {
+  # Remove double or single quotes.
+  local id=$(echo "${1}" | xargs echo)
+  local get_path="pipeline/${id}"
+  local get_result=$(get "${get_path}")
+  __=$(echo "${get_result}" | jq -r .number)
+}
+
+# https://circleci.com/docs/2.0/workflows/#states
+# v2 workflow api tells status, but not the pipeline api.
+fetch_workflow_status() {
+  # Remove double or single quotes.
+  local id=$(echo "${1}" | xargs echo)
+  local get_path="pipeline/${id}/workflow"
+  local get_result=$(get "${get_path}")
+  local workflow_summary=$(echo "${get_result}" | jq ".items[] | {name: .name, id: .id, status: .status, pipeline_number: .pipeline_number}")
+  printf "${workflow_summary}\n"
+  # Rerunning a failed workflow produces a nested datetime sorted array. Get the status of latest run (first array element).
+  __=$(echo "${get_result}" | jq -r "first(.items[]) | select(.name==\"${workflow_name}\") | .status | @sh")
+}
+
+fetch_this_pipeline_id() {
+  local get_path="workflow/${CIRCLE_WORKFLOW_ID}"
+  local get_result=$(get "${get_path}")
+  __=$(echo "${get_result}" | jq -r '.pipeline_number')
+}
+
+# Skip checking once condition is met.
+should_skip() {
+  if [[ "$CIRCLE_BRANCH" != "master" ]]; then
+    printf "Not on master branch. Skip checking.\n"
+    return 0
+  fi
+
+  # Get the pipeline id of this job.
+  fetch_this_pipeline_id
+  this_pipeline_id=$__
+  if [[ "$this_pipeline_id" == $1 ]]; then
+    # Don't check this pipeline if this pipeline is itself.
+    printf '%s\n' "Not waiting on myself (pipeline id: \"${this_pipeline_id}\"). Skip checking.\n" >&2
+    return 0
+  fi
+
+  return 1
+}
+
+#********************
+# RUNNING
+# *******************
+
+printf "\n"
+check_circleci_token
+
+# Get recent pipelines.
+fetch_pipeline_ids
+pipeline_ids=$(jq '. | .id' ${pipeline_json} | jq -r @sh)
+
+# Check workflow status in each pipeline. Wait while workflow status is running or failing.
+wait="30s"
+IFS=$'\n'
+for id in ${pipeline_ids}; do
```

Comment:
I'm thinking through whether this works - specifically about the fact that we're grabbing pipeline IDs once at an early point in time, and then repeatedly polling the same set (also, what pipeline IDs will be in this set).

1. are pipeline IDs incremental? Are we making some kind of assumption based on this, or else can we look at some metadata returned by fetch_pipelines which indicates the pipeline submission time?

2. Are we somehow ignoring pipelines that were launched after us? Feels like you could get into a situation where two pipelines are sitting in this sleep loop waiting for eachother to finish. When considering pipeline IDs to block on, it would be ideal to first filter down to only pipelines that were submitting before ourselves.

Reply:
> So it is true that script considers only pipelines that were submitted before.

It only considers pipelines that were submitted before _this code runs_, which may be different from when this pipeline was submitted. Circle jobs can be submitted in parallel, so I don't think this is making a safe assumption. I think probably you would want to filter out any pipelines which come after ourselves - ideally using the creation time.

Reply:
Maybe I'm wrong about this, the `/pipeline` response does not include any pipeline which were submitted after ourselves. I have not seen that during my testing.

If we cannot use a personal token to work with `v2` APIs, how about using `v1` APIs? `v1` APIs supposedly still work with project token. But, it's a matter of time, v1 APIs will not be supported anymore by CircleCI in future. We can choose to ignore that for now.

Reply:
From CircleCI [doc](https://circleci.com/docs/2.0/api-intro/):

```
API v2 includes several powerful features (e.g. support for pipelines and pipeline parameters) that are unavailable in API v1.1. It is recommended that CircleCI Cloud users migrate their scripts to API v2 stable endpoints as soon as possible.

CircleCI API v1.1 and portions of API v2 are supported and generally available. CircleCI expects to eventually End-Of-Life (EOL) API v1.1 in favor of API v2 as more API v2 endpoints are announced as stable. Further guidance on when CircleCI API v1.1 will be discontinued will be communicated at a future date.
```